lib/membrane/rtp/twcc_sender/congestion_control.ex

```diff
@@ -0,0 +1,386 @@
+    %__MODULE__{
+      r_hats: r_hats,
+      a_hat: prev_a_hat,
+      last_bandwidth_increase_ts: last_bandwidth_increase_ts,
+      packet_received_sizes: packet_received_sizes
+    } = cc
+
+    packet_received_interval_ms =
+      Time.to_milliseconds(packet_received_interval_end - packet_received_interval_start)
+
+    r_hat = 1 / (packet_received_interval_ms / 1000) * packet_received_sizes
+
+    now = Time.vm_time()
+    last_bandwidth_increase_ts = last_bandwidth_increase_ts || now
+    time_since_last_update_ms = Time.to_milliseconds(now - last_bandwidth_increase_ts)
+
+    a_hat =
+      case bitrate_increase_mode(r_hat, cc) do
+        :multiplicative ->
+          eta = :math.pow(1.08, min(time_since_last_update_ms / 1000, 1))
+          eta * prev_a_hat
+
+        :additive ->
+          response_time_ms = 100 + Time.to_milliseconds(rtt)
+          alpha = 0.5 * min(time_since_last_update_ms / response_time_ms, 1)
+          expected_packet_size_bits = Enum.sum(packet_sizes) / length(packet_sizes)
+          prev_a_hat + max(1000, alpha * expected_packet_size_bits)
+      end
+
+    a_hat = min(1.5 * r_hat, a_hat)
+
+    %__MODULE__{
+      cc
+      | a_hat: a_hat,
+        r_hats: Enum.take([r_hat | r_hats], @last_receive_bandwidth_probe_size),
+        last_bandwidth_increase_ts: now,
+        packet_received_interval_end: nil,
+        packet_received_interval_start: nil,
+        packet_received_sizes: 0
+    }
+  end
+
+  defp update_receiver_bandwidth(cc, _packet_sizes, _rtt), do: cc
+
+  defp update_sender_bandwidth(%__MODULE__{as_hat: as_hat, a_hat: a_hat} = cc, receive_deltas) do
+    lost = Enum.count(receive_deltas, &(&1 == :not_received))
+    loss_ratio = lost / length(receive_deltas)
+
+    as_hat =
+      cond do
+        loss_ratio < 0.02 -> 1.05 * as_hat
+        loss_ratio > 0.1 -> as_hat * (1 - 0.5 * loss_ratio)
+        true -> as_hat
+      end
+
+    %__MODULE__{cc | as_hat: min(as_hat, 1.5 * a_hat)}
+  end
+
+  defp bitrate_increase_mode(_r_hat, %__MODULE__{r_hats: prev_r_hats})
+       when length(prev_r_hats) < @last_receive_bandwidth_probe_size,
+       do: :multiplicative
+
+  defp bitrate_increase_mode(r_hat, %__MODULE__{r_hats: prev_r_hats}) do
+    exp_average = exponential_moving_average(@ema_smoothing_factor, prev_r_hats)
+    std_dev = std_dev(prev_r_hats)
+
+    if abs(r_hat - exp_average) <= 3 * std_dev do
+      :additive
+    else
+      :multiplicative
+    end
+  end
+
+  defp exponential_moving_average(_alpha, []), do: 0
+
+  defp exponential_moving_average(alpha, [latest_observation | older_observations]) do
+    alpha * latest_observation +
+      (1 - alpha) * exponential_moving_average(alpha, older_observations)
+  end
+
+  defp std_dev(observations) when observations != [] do
+    n_obs = length(observations)
+    mean = Enum.sum(observations) / n_obs
+
+    observations
+    |> Enum.reduce(0, fn obs, acc -> acc + (obs - mean) * (obs - mean) end)
```

Comment:
Not sure about this one – there's no `Enum.sum/2` function, so we would need to use `Enum.map/2` first. Also, dialyzer complains at `** 2` here as it can't infer the result type for `**` (wants `** 2.0`).

Reply:
Multiplying seems to be twice as performant
```
Name               ips        average  deviation         median         99th %
a * a           6.97 K      143.52 μs     ±3.12%      141.99 μs      162.99 μs
** int          3.41 K      293.19 μs     ±1.34%      292.99 μs      310.99 μs
** float        2.74 K      365.22 μs     ±3.96%      362.99 μs      392.99 μs

Comparison: 
a * a           6.97 K
** int          3.41 K - 2.04x slower +149.68 μs
** float        2.74 K - 2.54x slower +221.70 μs
```